src/libraries/System.Runtime/tests/System.IO.FileSystem.Tests/RandomAccess/NonSeekable.cs

```diff
@@ -0,0 +1,429 @@
+
+                await write;
+            }
+        }
+
+        [Fact]
+        public async Task MultipleBuffersAreSupported_AsyncRead_SyncWrite()
+        {
+            (SafeFileHandle readHandle, SafeFileHandle writeHandle) = await GetNamedPipeHandlesAsync();
+
+            using (readHandle)
+            using (writeHandle)
+            {
+                Memory<byte>[] writableVectors = GenerateVectors(VectorCount, BufferSize);
+                ValueTask<long> read = RandomAccess.ReadAsync(readHandle, writableVectors, fileOffset: 456);
+
+                byte[] content = RandomNumberGenerator.GetBytes(VectorsByteCount);
+                RandomAccess.Write(writeHandle, content, fileOffset: 123);
+
+                int bytesRead = (int)await read;
+                Assert.InRange(bytesRead, 1, VectorsByteCount);
+                Assert.Equal(content.Take(bytesRead), writableVectors.SelectMany(vector => vector.ToArray()).Take(bytesRead));
+            }
+        }
+
+        private static ReadOnlyMemory<byte>[] GenerateReadOnlyVectors(int vectorCount, int bufferSize)
+            => Enumerable.Range(0, vectorCount).Select(_ => new ReadOnlyMemory<byte>(RandomNumberGenerator.GetBytes(bufferSize))).ToArray();
+
+        private static Memory<byte>[] GenerateVectors(int vectorCount, int bufferSize)
+            => Enumerable.Range(0, vectorCount).Select(_ => new Memory<byte>(RandomNumberGenerator.GetBytes(bufferSize))).ToArray();
+
+        private static SafeFileHandle GetFileHandle(PipeStream pipeStream, bool isAsync)
+        {
+            var serverHandle = new SafeFileHandle(pipeStream.SafePipeHandle.DangerousGetHandle(), ownsHandle: true);
+
+            try
+            {
+                if (OperatingSystem.IsWindows() && isAsync)
+                {
+                    // Currently it's impossible to duplicate an async safe handle that has already been bound to Thread Pool: https://github.com/dotnet/runtime/issues/28585
+                    FileThreadPoolBoundHandle(serverHandle) = PipeThreadPoolBoundHandle(pipeStream);
+                }
+
+                return serverHandle;
+            }
+            finally
+            {
+                pipeStream.SafePipeHandle.SetHandleAsInvalid();
+            }
+
+            [UnsafeAccessor(UnsafeAccessorKind.Field, Name = "_threadPoolBinding")]
+            extern static ref ThreadPoolBoundHandle PipeThreadPoolBoundHandle(PipeStream @this);
+
+            [UnsafeAccessor(UnsafeAccessorKind.Field, Name = "<ThreadPoolBinding>k__BackingField")]
+            extern static ref ThreadPoolBoundHandle FileThreadPoolBoundHandle(SafeFileHandle @this);
+        }
+
+        private static void ReadExactly(SafeFileHandle readHandle, Span<byte> buffer, int expectedByteCount)
+        {
+            int bytesRead = 0;
+            do
+            {
+                bytesRead += RandomAccess.Read(readHandle, buffer.Slice(bytesRead), fileOffset: 0); // fileOffset NOT set to bytesRead on purpose
```

Comment:
Would it be good to throw or break out of the loop if RandomAccess.Read returns 0 to avoid a potential infinite loop?